### PR TITLE
Add proper clang-format to pre-commit

### PR DESCRIPTION
**Pull request by Skylion007**
## Motivation and Context
* Makes clang-format a proper pre-commit hook using a new pip package we created for PyBind11
* Automatically handles clang-format versioning and dependencies with the new hook

## Types of changes

- [X] Docs change / refactoring / dependency upgrade

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -102,15 +102,11 @@ repos:
...
     hooks:
-    - id: clang-format
-      name: Run clang-format
-      exclude: src/deps
-      entry: clang-format -i -style=file
-      types: [text]
-      files: '.*\.(cpp|h|hpp|cu|cuh)$'
-      language: system
+    -   id: clang-format
+        types_or: [c++, c, c#, cuda] # also run on CUDA
```
Comment (jturner65) [inline .pre-commit-config.yaml:109]: c# ? :partying_face:

**Review by jturner65 — APPROVED** [review]
LGTM

File: .pre-commit-config.yaml (hunk 1 above)
Comment (Skylion007) [reply .pre-commit-config.yaml:109]: It's supported, so why not. Main issue is that the OG hook doesn't support CUDA as a type, although I am working on getting it fixed. 

Other supported languages include JSN btw.

Comment (Skylion007) [reply .pre-commit-config.yaml:109]: The CUDA type_or is now merged in, but won't be updated until a new release is made of the hook.

**Pull request merged**